### ISSUE TITLE
Space betwee type annoation and [] or ... not verifiable for Google Style

### ISSUE THREAD

**Issue by Vampire** [checkstyle/checkstyle#6707]
https://checkstyle.org/styleguides/google-java-style-20180523/javaguide.html

`$ cat config.xml`
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="NoWhitespaceBefore"/>
    </module>
</module>
```
`$ cat Test.java`
```java
public class Test {

    @C int @A [] @B [] f;

    public void foo(final char @NonNull [] param) {}

    void test1(String @NonNull ... param) {} // violation

    void test2(String @NonNull... param) {} // OK

}
```
```
$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar /var/tmp/checkstyle-8.31-SNAPSHOT-all.jar \
          -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:7:32: '...' is preceded with whitespace. [NoWhitespaceBefore]
Audit done.
Checkstyle ends with 1 errors.
```

--------------------

> 4.6.2 Horizontal whitespace
........
Between a type annotation and [] or ....

Latest version of Google Style Guide mandates that there is a single space between a type annotation and `[]` or `...`.
There is currently no rule to check this.

Actually it is even worse, the [current config](https://checkstyle.org/google_style.html#a4.6.2) (`NoWhitespaceBefore` with `ELLIPSIS` token) will not only not allow to verify the Google style guide, but report violations where it shouldn't as it does not allow spaces in front of ellipsis tokens.

**Comment by HuGanghui** [comment]
I am on it

**Comment by rnveach** [comment]
@romani Please confirm how this issue should be fixed. ELLIPSIS token was added by user request as seen in PR. Google doc specifies between annotation and `...`. I think we enforce no space before only if there is no annotation.

**Comment by romani** [comment]
it is really sad that I for some reason approved issue that is not followed issue template .... right now it is not very clear what cases we try to resolve.

@HuGanghui , please do me a favor and make a comment in this issue and provide CLI output of current behavior and with `type annotation and [] or ....`.

**Comment by HuGanghui** [comment]
@romani ok,~
Output of configure the check:
$ cat config.xml
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="NoWhitespaceBefore"/>
    </module>
</module>
```
$ cat Test.java
```java
public class Test {

    @C int @A [] @B [] f;

    public void foo(final char @NonNull [] param) {}

    void test1(String @NonNull ... param) {} // violation

    void test2(String @NonNull... param) {} // OK

}
```
$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar /var/tmp/checkstyle-8.31-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:7:32: '...' is preceded with whitespace. [NoWhitespaceBefore]
Audit done.
Checkstyle ends with 1 errors.

**Comment by romani** [comment]
@HuGanghui , thanks a lot, I moved your code example to issue, please share PR with fix.

**Comment by romani** [comment]
```
$ cat Test.java 
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

public class Test {

  @Target(ElementType.TYPE_USE)
  @interface NonNull {
  }

  @NonNull int @NonNull[] @NonNull[] fiel1;
  @NonNull int @NonNull [] @NonNull [] field2;
  //@NonNull int @NonNull ... field3; // non-compilable
  //@NonNull int @NonNull... field4; // non-compilable


  public void foo2(final char[] param) {
  }

  public void foo1(final char [] param) {
  }

  public void foo3(final char @NonNull[] param) {
  }

  public void foo4(final char @NonNull [] param) {
  }

  void test4(String... param) {
  }

  void test3(String ... param) { // violation on 8.32
  }

  void test2(String @NonNull... param) {
  }

  void test1(String @NonNull ... param) { // violation on 8.32
  }
}

$ java -jar /var/tmp/checkstyle-8.32-all.jar -c /google_checks.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:31:21: '...' is preceded with whitespace. [NoWhitespaceBefore]
[WARN] /var/tmp/Test.java:37:30: '...' is preceded with whitespace. [NoWhitespaceBefore]
Audit done.

```

**Comment by strkkk** [comment]
Fix is merged
